V2 manifest save tarsum

dmcgowan opened the pull request:
Stores checksum in a separate file to avoid modifying the existing json file.  Storing this file allows for subsequent calls to create the manifest to use the cached value, avoiding unnecessary checksum calculations.  The checksum should not need to be recalculated if it was calculated on commit.

fixes #10269
fixes #10432

icecrime commented:
*If* [this](https://github.com/docker/docker/pull/10322#issuecomment-71743746) is correct, than I believe this can wait for 1.5.1.

dmcgowan commented:
@vbatts given the manifest json is already being altered, would you still oppose overwriting the layer json file for this change?  This is what my original commit did and given compatibility with tools relying on it for detached signature is already broken, seems pointless to introduce another file.

dmcgowan commented:
Updated to include fix for #10432. This will now ignore the value in the img.Checksum since it may differ from the checksum produced by the diffdriver for the registry.

@vbatts ignore my previous comment, separating out the checksum into a different file is necessary to fix an issue with computing the tarsum on import/pull/load before it is ever extracted.

crosbymichael commented:
much better ;)

crosbymichael commented:
What about deleting the Checksum field on the image now that it's not used?

dmcgowan commented:
The Checksum field is not new to this release.  If we want to get rid of it I would prefer it in a separate change to keep this one focused on push tarsum.

crosbymichael commented:
LGTM

icecrime commented on graph/manifest.go:
```diff
@@ -108,6 +111,11 @@ func (s *TagStore) newManifest(localName, remoteName, tag string) ([]byte, error
+
+			// Save checksum value
+			if err := layer.SaveCheckSum(s.graph.ImageRoot(layer.ID), checksum); err != nil {
+				return nil, fmt.Errorf("Error storing image checksum: %s", err)
```
`SaveCheckSum` already returns a rich error message. On error, this will output:

    Error storing image checksum: Error storing checksum in foo/bar/checksum: some error message

dmcgowan committed:
Store tar checksum in separate file

Fixes #10432

Signed-off-by: Derek McGowan <derek@mcgstyle.net> (github: dmcgowan)

dmcgowan commented:
@icecrime updated error message

Commit:
Remove Checksum field from image.Image struct

The checksum is now being stored in a separate file beside the image
JSON file.

Docker-DCO-1.1-Signed-off-by: Josh Hawn <josh.hawn@docker.com> (github: jlhawn)

dmcgowan commented:
@crosbymichael added a commit from @jlhawn to remove the now unused Checksum field.  He was able to confirm it is not needed anywhere else.

crosbymichael commented:
Thanks!

LGTM

crosbymichael commented:
@dmcgowan are there any issues that this fixes around performance or something?

dmcgowan commented:
Retroactively fixes problems created by #10432 and fixes #10269

The #10269 issue is currently in 1.5.1 but I think it would be good to fix for 1.5.0 since it will now effect images built by 1.5.0.

crosbymichael commented:
@dmcgowan please add the Fixes #xxxxx to your pr descriptions so they are alto closed on merge

dmcgowan commented:
@crosbymichael done

crosbymichael commented:
ping @tiborvass @jfrazelle

jfrazelle commented:
LGTM

icecrime commented:
LGTM!

Handing off to @tiborvass to let him finish his own tests and merge when appropriate.

tiborvass commented:
LGTM
